### ISSUE TITLE
jwt.verify() not truly async.

### ISSUE THREAD

**Issue by vicary** [auth0/node-jsonwebtoken#511]
Despite the verify action is in fact synchronize, the callback API does not.

It is specifically this line that [released Zalgo](https://medium.com/@bluepnume/intentionally-unleashing-zalgo-with-promises-ab3f63ead2fd).

https://github.com/auth0/node-jsonwebtoken/blob/5a7fa23c0b4ac6c25304dab8767ef840b43a0eca/verify.js#L24

Which can be fixed by wrapping it as such

```js
    done = function() {
      setImmediate(callback, arguments);
    };
```

**Comment by ziluvatar** [comment]
That behavior (with process.nextTick() though) was removed in this PR: https://github.com/auth0/node-jsonwebtoken/pull/302

I will leave the issue open in case the community has strong opinions about re-adding it or keep it as is. I understand the reasons, however, the pain to the event loop was already done while performing the verification synchronously, even doing that is "lying" to the consumer.

There was another discussion to remove the callback-ish style to that function, we didn't do it and now we support to get the secret/key asynchronously which gives sense to the callback now.

**Comment by vicary** [comment]
Forgive my words if the opinions is too strong, but to the best of my knowledge the consensus is already there for decades.

The nature of an asynchronous API should behave like it means, implementation should be hidden behind the API even if it is a so-called lie.

The API is the contract of behavior between the library and its users, and should be followed the moment it goes public as **(Asynchronous) ...**.

IMO #302 should in fact use the **(Synchronous) ...** version without API for the desired *performance* gain, and such performance differences should be documented instead of breaking the API.

**Comment by vicary** [comment]
@MitMaro thanks.

I don't have much on the callback/promise/async controversy because of the disproportionate gain of productivity, the difference in performance is tiny; I like the inline error catching in async/await though.

But the breaking of the current API should be a hotfix instead of next major, there are potential risks for people using them in production.

**Comment by mcmar** [comment]
Hey, I just want to point out that if you ever re-add code to make this async (which you should do), you should also really switch from `process.nextTick` to `setImmediate`. You can read about the differences here: https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/

TL;DR `process.nextTick` doesn't allow the event loop time to handle things like I/O buffering before you run the verification function. Since cryptographic functions can take more than a few processor cycles to execute, you really do want to allow node to give itself a chance to handle its I/O before starting on your big chunk of work. Otherwise this library will negatively impact I/O perf much more than is necessary.

**Comment by nhhockeyplayer** [comment]
yeah this it wiped out all my elegant graceful symmetrically patterned async/await logic
whoever is pushing this stuff really needs to stop
I see it in the field even today by juniors claiming hail for closure when they never modeled a single noun or verb

callbacks had their day in DECWindows back in 1989 
but its over finally let it die
and I cant stand getting grilled on interviews about 'what is a callback' I hangup the phone
glad I never touched callbacks in javascript only for a promise based rest service
same thing with closures a1 was enough

but this is what happens when these hip scripters and python pushers punch a hole thru the middle tier

nice thing about koa2 is we can take their little nifty jwtVerify code and stuff it atop the koa2 middlware stack

**Comment by cyrilchapon** [comment]
But, is there a reason such a heavy processing is done synchronously in the first place ?

The fact you remove the fake-callback-still-synchronous strategy is not really addressing that issue.

**Comment by mcmar** [comment]
@cyrilchapon Everything in javascript is inherently single-threaded. Doing it "asynchronously" won't allow any other code to execute simultaneously. It'll just return after at least 1 tick of the event loop (`process.nextTick`) and it might also give the event loop time to handle I/O streams (`setImmediate`)

**Comment by mcmar** [comment]
We *might* be able to use something like [threads.js](https://github.com/andywer/threads.js) to do this in a real multi-threaded manner, which would be fun and sexy for everyone. But it also might reduce compatibility and/or require pseudo-polyfills like [webworker-fallback](https://github.com/andywer/webworker-fallback)

**Comment by mohamedabdelmaksod87** [comment]
I believe the verify method is synchronous in nature and if we considered it a computationally expensive process we may use the node.js worker threads module to run it in parallel out of the main thread from the performance point of view.
